File: service/app/src/main/java/io/miragon/blueprint/application/service/GetLeasingApplicationService.java
package io.miragon.blueprint.application.service;

import io.miragon.blueprint.application.port.inbound.GetLeasingApplicationQuery;
import io.miragon.blueprint.application.port.outbound.BikePortfolioRepository;
import io.miragon.blueprint.application.port.outbound.LeasingApplicationRepository;
import io.miragon.blueprint.domain.bike.Bike;
import io.miragon.blueprint.domain.leasing.ApplicationId;
import java.util.Optional;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(readOnly = true)
public class GetLeasingApplicationService implements GetLeasingApplicationQuery {

    private final LeasingApplicationRepository repository;
    private final BikePortfolioRepository bikePortfolio;

    public GetLeasingApplicationService(
            LeasingApplicationRepository repository,
            BikePortfolioRepository bikePortfolio) {
        this.repository = repository;
        this.bikePortfolio = bikePortfolio;
    }

    @Override
    public Optional<GetLeasingApplicationQuery.Result> byId(ApplicationId id) {
        return repository.findById(id)
                .map(application -> new GetLeasingApplicationQuery.Result(
                        application,
                        bikePortfolio.findByBikeId(application.bikeId()).map(Bike::model).orElse(null)));
    }
}

File: service/app/src/main/kotlin/io/miragon/blueprint/application/service/GetLeasingApplicationService.kt
package io.miragon.blueprint.application.service

import io.miragon.blueprint.application.port.inbound.GetLeasingApplicationQuery
import io.miragon.blueprint.application.port.outbound.BikePortfolioRepository
import io.miragon.blueprint.application.port.outbound.LeasingApplicationRepository
import io.miragon.blueprint.domain.leasing.ApplicationId
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class GetLeasingApplicationService(
    private val repository: LeasingApplicationRepository,
    private val bikePortfolio: BikePortfolioRepository,
) : GetLeasingApplicationQuery {

    override fun byId(id: ApplicationId): GetLeasingApplicationQuery.Result? {
        val application = repository.findById(id) ?: return null
        val bike = bikePortfolio.findByBikeId(application.bikeId)
        return GetLeasingApplicationQuery.Result(application, bike?.model)
    }
}
